### ISSUE TITLE
[Tune (might be Core)] Application error with no informative error messages

### ISSUE THREAD

**Issue by YuhangSong** [ray-project/ray#33991]
### What happened + What you expected to happen

Thank you all for making/contributing to this amazing project and making it publicly available first, I have been using ray for a long time, it's been very helpful!

I am using the function API of ray Tune [here](https://docs.ray.io/en/latest/tune/key-concepts.html#ray-tune-trainables). In the function, called my_function, I am using [lm_eval](https://github.com/EleutherAI/lm-evaluation-harness) to evaluate a 70m size language model.

The function takes around 20 minutes to finish. 

I am using ray job + cluster.

If I submit this function to the ray cluster without using ray, i.e., `ray job submit -- python my_function.py`, it runs fine and log
```
------------------------------------------
Job 'raysubmit_mabQummiVibBPmWW' succeeded
------------------------------------------
```

If I submit my script, `main.py`, to a ray cluster and inside the `main.py`, there is `ray.init(...)` and `from my_function import my_function; ray.Tuner(my_function, ...)`

When run with `ray.init(local_mode=True)`, the function finishes successfully and returns correct results, but logged the following error messages:
```
== Status ==
Current time: 2023-04-01 09:25:23 (running for 00:23:01.88)
Memory usage on this node: 3.3/30.9 GiB 
Using FIFO scheduling algorithm.
Resources requested: 0/8 CPUs, 0/1 GPUs, 0.0/18.35 GiB heap, 0.0/9.18 GiB objects (0.0/1.0 accelerator_type:T4)
Result logdir: /root/ray_results/transformer_evaluate
Number of trials: 1/1 (1 TERMINATED)
+-------------------+------------+-------------------+------------------------+------------------------+------------------------+------------------------+------------------------+------------------------+--------+------------------+-------------+--------------+-------+
| Trial name        | status     | loc               | from_pretrained/revi   | from_pretrained/size   | model_config/lmatmul   |   model_config/lmatmul | ...l_config/lmatmul/   | ..._config/lmatmul/k   |   iter |   total time (s) |         ppl |   ppl_stderr |   acc |
|                   |            |                   | sion                   |                        | /enable                |        /kwargs/abs_min | kwargs/below_abs_min   | wargs/is_sot_product   |        |                  |             |              |       |
|-------------------+------------+-------------------+------------------------+------------------------+------------------------+------------------------+------------------------+------------------------+--------+------------------+-------------+--------------+-------|
| train_e9b4f_00000 | TERMINATED | 172.31.19.29:3760 | step13000              | 70m                    | True                   |                      0 | zero                   | True                   |      1 |          1375.69 | 9.87902e+07 |  1.03755e+07 |     0 |
+-------------------+------------+-------------------+------------------------+------------------------+------------------------+------------------------+------------------------+------------------------+--------+------------------+-------------+--------------+-------+


2023-04-01 09:25:23,871 INFO tune.py:798 -- Total run time: 1381.98 seconds (1375.95 seconds for the tuning loop).

Segmentation fault (core dumped)

---------------------------------------
Job 'raysubmit_2it441aF9EUYf8hv' failed
---------------------------------------

Status message: Job failed due to an application error, last available logs (truncated to 20,000 chars):
+-------------------+------------+-------------------+------------------------+------------------------+------------------------+------------------------+------------------------+------------------------+--------+------------------+-------------+--------------+-------+


2023-04-01 09:25:23,871 INFO tune.py:798 -- Total run time: 1381.98 seconds (1375.95 seconds for the tuning loop).

Segmentation fault (core dumped)
```

When run **NOT** in local_model, the job is "killed" around after around 4 minutes, so 1/5 of the complete process (the exact internal steps that the function gets killed is different slightly from run to run), the logged error message is:
```
== Status ==
Current time: 2023-04-01 09:33:36 (running for 00:05:05.35)
Memory usage on this node: 3.9/30.9 GiB 
Using FIFO scheduling algorithm.
Resources requested: 8.0/8 CPUs, 1.0/1 GPUs, 0.0/18.35 GiB heap, 0.0/9.18 GiB objects (0.0/1.0 accelerator_type:T4)
Result logdir: /root/ray_results/transformer_evaluate
Number of trials: 1/1 (1 RUNNING)
+-------------------+----------+-------------------+------------------------+------------------------+------------------------+------------------------+------------------------+------------------------+
| Trial name        | status   | loc               | from_pretrained/revi   | from_pretrained/size   | model_config/lmatmul   |   model_config/lmatmul | ...l_config/lmatmul/   | ..._config/lmatmul/k   |
|                   |          |                   | sion                   |                        | /enable                |        /kwargs/abs_min | kwargs/below_abs_min   | wargs/is_sot_product   |
|-------------------+----------+-------------------+------------------------+------------------------+------------------------+------------------------+------------------------+------------------------|
| train_91086_00000 | RUNNING  | 172.31.19.29:5613 | step13000              | 70m                    | True                   |                      0 | zero                   | True                   |
+-------------------+----------+-------------------+------------------------+------------------------+------------------------+------------------------+------------------------+------------------------+



---------------------------------------
Job 'raysubmit_RZ9wYCKJVjrGLS15' failed
---------------------------------------

Status message: Job failed due to an application error, last available logs (truncated to 20,000 chars):
Result logdir: /root/ray_results/transformer_evaluate
Number of trials: 1/1 (1 RUNNING)
+-------------------+----------+-------------------+------------------------+------------------------+------------------------+------------------------+------------------------+------------------------+
| Trial name        | status   | loc               | from_pretrained/revi   | from_pretrained/size   | model_config/lmatmul   |   model_config/lmatmul | ...l_config/lmatmul/   | ..._config/lmatmul/k   |
|                   |          |                   | sion                   |                        | /enable                |        /kwargs/abs_min | kwargs/below_abs_min   | wargs/is_sot_product   |
|-------------------+----------+-------------------+------------------------+------------------------+------------------------+------------------------+------------------------+------------------------|
| train_91086_00000 | RUNNING  | 172.31.19.29:5613 | step13000              | 70m                    | True                   |                      0 | zero                   | True                   |
+-------------------+----------+-------------------+------------------------+------------------------+------------------------+------------------------+------------------------+------------------------+



```
And there are no error messages below, just normal logs.

But without further indication of what the error is.

I looked at all the log files on the cluster, I find the following messages:
```
[2023-04-01 08:15:41,551 I 750 783] core_worker.cc:3406: Force kill actor request has received. exiting immediately... The actor is dead because all references to the actor were removed.238[2023-04-01 08:15:41,551 W 750 783] core_worker.cc:778: Force exit the process.  Details: Worker exits because the actor is killed. The actor is dead because all references to the actor were removed.239[2023-04-01 08:15:41,552 I 750 783] core_worker.cc:698: Disconnecting to the raylet.240[2023-04-01 08:15:41,552 I 750 783] raylet_client.cc:163: RayletClient::Disconnect, exit_type=INTENDED_SYSTEM_EXIT, exit_detail=Worker exits because the actor is killed. The actor is dead because all references to the actor were removed., has creation_task_exception_pb_bytes=0
```

In another one
```
239[2023-04-01 08:44:10,897 I 2107 2150] core_worker.cc:711: Exit signal received, this process will exit after all outstanding tasks have finished, exit_type=INTENDED_USER_EXIT, detail=Worker exits by an user request. exit_actor() is called.240[2023-04-01 08:44:10,898 I 2107 2107] core_worker.cc:698: Disconnecting to the raylet.241[2023-04-01 08:44:10,898 I 2107 2107] raylet_client.cc:163: RayletClient::Disconnect, exit_type=INTENDED_USER_EXIT, exit_detail=Worker exits by an user request. exit_actor() is called., has creation_task_exception_pb_bytes=0242[2023-04-01 08:44:10,898 I 2107 2107] core_worker.cc:621: Shutting down a core worker.243[2023-04-01 08:44:10,898 I 2107 2107] concurrency_group_manager.cc:83: Default executor is joining. If the 'Default executor is joined.' message is not printed after this, the worker is probably hanging because the actor task is running an infinite loop.244[2023-04-01 08:44:10,898 I 2107 2107] concurrency_group_manager.cc:87: Default executor is joined.245[2023-04-01 08:44:10,898 I 2107 2107] task_event_buffer.cc:160: Shutting down TaskEventBuffer.246[2023-04-01 08:44:10,898 I 2107 2141] task_event_buffer.cc:142: Task event buffer io service stopped.247[2023-04-01 08:44:10,898 I 2107 2107] core_worker.cc:647: Disconnecting a GCS client.248[2023-04-01 08:44:10,898 I 2107 2107] core_worker.cc:651: Waiting for joining a core worker io thread. If it hangs here, there might be deadlock or a high load in the core worker io service.249[2023-04-01 08:44:10,898 I 2107 2135] core_worker.cc:798: Core worker main io service stopped.250[2023-04-01 08:44:10,902 I 2107 2107] core_worker.cc:664: Core worker ready to be deallocated.251[2023-04-01 08:44:10,902 I 2107 2107] core_worker_process.cc:243: Task execution loop terminated. Removing the global worker.252[2023-04-01 08:44:10,902 I 2107 2107] core_worker.cc:612: Core worker is destructed253[2023-04-01 08:44:11,128 I 2107 2107] core_worker_process.cc:147: Destructing CoreWorkerProcessImpl. pid: 2107254[2023-04-01 08:44:11,128 I 2107 2107] io_service_pool.cc:47: IOServicePool is stopped.255
```

Still, there is no more specific message about why this happens. I suspect it is because some of the resource limits are being hit. Or ray core has some bug handling garbage cleaning that wrongly thought all the objects in the function are not referred to anymore? I am saying it is "wrong" because when runs in the local model, the function finishes successfully (tried several times, and all finished successfully), also when run `my_function.py` directly without `ray.init` and `ray.Tuner` it runs fine.

I asked ChatGPT-4 about this and it did not provide more useful information.

Any instructions on where to look to resolve this bug will be appreciated. Besides, I think, as said above, it might indicate an improvement of ray's logging and tracking error needs to be improved, or there is a core bug of wrongly assuming all objects are not referred.

### Versions / Dependencies

- Python 3.8
- Ray 2.3.0
- Ubuntu

### Reproduction script

I will try to produce an independent reproduction script, but before I do that, someone might provide a hint on where to look for the problem and it can already be solved.

### Issue Severity

High: It blocks me from completing my task.

**Comment by rkooo567** [comment]
@Yard1 can you have the initial investigation and let us know if this is the core issue?

**Comment by YuhangSong** [comment]
An update, it does not seem to be a core issue, I replaced my docker image with another one I used more frequently and it worked. So the issue is with the docker image I built, although from the above-described steps for me to look into this problem, it seems to have ruled out the problem of the docker:

- It runs fine when you have run with local mode

I think this might indicate that there should be some improvements to the log or error messages. But since it is now not blocking, feel free to close it.

@rkooo567 Thx for pinning a message on this though!
